CMakeLists.txt

```
@@ -1,5 +1,273 @@
-cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.16)
 
-project(usb-1.0 C)
+# define the source root
+set(LIBUSB_ROOT "libusb/libusb")
 
-message(FATAL_ERROR "Finish me!")
+# Get the version information from version.h ignoring the nano version as it appears in version_nano.h and so we need it?
+file(READ "${LIBUSB_ROOT}/version.h" VERSIONHEADERDATA)
+string(REGEX MATCH "#define LIBUSB_MAJOR ([0-9]*)" _ ${VERSIONHEADERDATA} )
+set(LIBUSB_VERSION_MAJOR ${CMAKE_MATCH_1})
+string(REGEX MATCH "#define LIBUSB_MINOR ([0-9]*)" _ ${VERSIONHEADERDATA} )
+set(LIBUSB_VERSION_MINOR ${CMAKE_MATCH_1})
+string(REGEX MATCH "#define LIBUSB_MICRO ([0-9]*)" _ ${VERSIONHEADERDATA} )
+set(LIBUSB_VERSION_MICRO ${CMAKE_MATCH_1})
+set(LIBUSB_VERSION "${LIBUSB_VERSION_MAJOR}.${LIBUSB_VERSION_MINOR}.${LIBUSB_VERSION_MICRO}")
+
+project(usb-1.0
+  DESCRIPTION "A cross-platform library to access USB devices"
+  VERSION ${LIBUSB_VERSION}
+  LANGUAGES C )
+
+if(MINGW)
+  # This is because CMake on mingw can incorrectly use response files for includes
+  # instead of passing them directly via -I<dir>
+  # Explicitly disable use of response files
+  set(CMAKE_C_USE_RESPONSE_FILE_FOR_INCLUDES OFF)
+endif()
+
+# This function generates all the local variables what end up getting written to config.
+# We use a function as any vars set in this context don't mess with the rest of the file.
+# e.g. Logging LIBUSB_ENABLE_LOGGING mapps to ENABLE_LOGGING in the config, keeps it clean
+function(GenerateConfigFile)
+    include(CheckIncludeFiles)
+    include(CheckFunctionExists)
+    include(CheckSymbolExists)
+    include(CheckStructHasMember)
+    include(CheckCCompilerFlag)
+
+    check_function_exists(clock_gettime             HAVE_CLOCK_GETTIME)
+    check_function_exists(pthread_condattr_setclock HAVE_PTHREAD_CONDATTR_SETCLOCK)
+    check_function_exists(pthread_setname_np        HAVE_PTHREAD_SETNAME_NP)
+    check_function_exists(pthread_threadid_np       HAVE_PTHREAD_THREADID_NP)
+    check_function_exists(eventfd                   HAVE_EVENTFD)
+    check_function_exists(pipe2                     HAVE_PIPE2)
+    check_function_exists(syslog                    HAVE_SYSLOG)
+
+    check_include_files(asm/types.h      HAVE_ASM_TYPES_H)
+    check_include_files(sys/eventfd.h    HAVE_EVENTFD)
+    check_include_files(string.h         HAVE_STRING_H)
+    check_include_files(sys/time.h       HAVE_SYS_TIME_H)
+
+    check_symbol_exists(timerfd_create  "sys/timerfd.h" HAVE_TIMERFD)
+    check_symbol_exists(nfds_t  "poll.h" HAVE_NFDS_T)
+
+    check_struct_has_member("struct timespec" tv_sec time.h HAVE_STRUCT_TIMESPEC)
+
+    if(HAVE_VISIBILITY)
+        set(DEFAULT_VISIBILITY "__attribute__((visibility(\"default\")))")
+    else()
+        set(DEFAULT_VISIBILITY "" )
+    endif()
+
+    # Set vars that will be written into the config file.
+    if(CMAKE_SYSTEM_NAME STREQUAL "Windows")
```

is there a particular reason here not to use the standard `WIN32` CMake variable?
```suggestion
    if(WIN32)
```